File: Solution/每日一题/2024.9.11/双指针.java
class Solution {
    public int maximizeWin(int[] prizePositions, int k) {
        int ans = 0;
        for (int l1 = 0, r1 = 0, l2 = 0, r2 = 0, firstMax = 0; r2 < prizePositions.length; r2++) {
            while (prizePositions[r2] - prizePositions[l2] > k) {
                l2++;
            }
            for (; r1 < l2; r1++) {
                while (prizePositions[r1] - prizePositions[l1] > k) {
                    l1++;
                }
                firstMax = Math.max(firstMax, r1 - l1 + 1);
            }
            ans = Math.max(ans, r2 - l2 + 1 + firstMax);
        }
        return ans;
    }
}

File: Solution/每日一题/2024.9.11/双指针.cpp
class Solution {
public:
    int maximizeWin(vector<int>& prizePositions, int k) {
        int n = prizePositions.size();
        vector<int> dp(n + 1);
        int ans = 0;
        for (int left = 0, right = 0; right < n; right++) {
            while (prizePositions[right] - prizePositions[left] > k) {
                left++;
            }
            ans = max(ans, right - left + 1 + dp[left]);
            dp[right + 1] = max(dp[right], right - left + 1);
        }
        return ans;
    }
};
